submariner-operator/diagnose-refactor-ocm.md

```diff
@@ -0,0 +1,137 @@
+# Refactor subctl diagonose for use with OCM
+
+## Summary
+
+Currently only way to run diagnostics on a Submariner installation is through `subctl` CLI. This requires user or admin
+to run it manually from a host. There is no way to run it through UI like OCM or deployment tools like helm charts etc.
+
+## Proposal
+
+This enhancement proposes a means to run `subctl diagnose` through a `SubmarinerDiagnose` Job. User or Admin can deploy
+this Job manually. In case of OCM, a new SubmarinerDiagnose Controller will be added to SubmarinerAddon which will deploy
+this Job, track its completion and fetch results once done.
+
+## Design Details
+
+* A new option `--output json` will be added to `subctl diagnose` to capture output in a json format instead of current
+ user readable text format.
+* `subctl diagnose` will also update some Prometheus metrices whenever it is run.
+* A `SubmarinerDiagnoseJob` Job will be created which will run `subctl diagnose` with relevant flags as `command`. This
+ job will run once to completion. This Job will always run with new option of `--output json`.
+* A new `SubmarinerDiagnose` CRD will be added to SubmarinerAddon. OCM UI will create this CR with relevant values
```

Comment:
Can a job has the same name as a CRD? `SubmarinerDiagnose`

Reply:
> Can a job has the same name as a CRD? `SubmarinerDiagnose`

Yes. Though open to suggestions for name here. going by convention, Job name should be `submariner-diagnose`. Maybe `submariner-diagnose-<randomized>` if we want to support multiple jobs running at a time. Note that most names are just placeholders for discussion and appreciate inputs on them.

Reply:
K8s can handle the name generation for us (`GenerateName` instead of `Name`). Having a fixed name makes everything simpler for now though!

Reply:
How will these objects be deleted?